src/slam_toolbox_common.cpp

```diff
@@ -63,49 +67,151 @@ void SlamToolbox::configure()
+/*****************************************************************************/
+{
+  RCLCPP_INFO(get_logger(), "Activating");
+  sst_->on_activate();
+  sstm_->on_activate();
+  pose_pub_->on_activate();
   closure_assistant_ =
     std::make_unique<loop_closure_assistant::LoopClosureAssistant>(
     shared_from_this(), smapper_->getMapper(), scan_holder_.get(),
     state_, processor_type_);
-  reprocessing_transform_.setIdentity();
 
-  double transform_publish_period = 0.05;
-  transform_publish_period =
-    this->declare_parameter("transform_publish_period",
-      transform_publish_period);
+  double transform_publish_period = this->get_parameter("transform_publish_period").as_double();
   threads_.push_back(std::make_unique<boost::thread>(
       boost::bind(&SlamToolbox::publishTransformLoop,
       this, transform_publish_period)));
   threads_.push_back(std::make_unique<boost::thread>(
       boost::bind(&SlamToolbox::publishVisualizations, this)));
+
+  scan_filter_ =
+    std::make_unique<tf2_ros::MessageFilter<sensor_msgs::msg::LaserScan>>(
+    *scan_filter_sub_, *tf_, odom_frame_, scan_queue_size_,
+    get_node_logging_interface(), get_node_clock_interface(),
+    tf2::durationFromSec(transform_timeout_.seconds()));
+  scan_filter_->registerCallback(
+    std::bind(&SlamToolbox::laserCallback, this, std::placeholders::_1));
+  return CallbackReturn::SUCCESS;
 }
 
 /*****************************************************************************/
-SlamToolbox::~SlamToolbox()
+CallbackReturn SlamToolbox::on_deactivate(const rclcpp_lifecycle::State &)
 /*****************************************************************************/
 {
+  RCLCPP_INFO(get_logger(), "Deactivating");
+  scan_filter_.reset();
   for (int i = 0; i != threads_.size(); i++) {
+    threads_[i]->interrupt();
     threads_[i]->join();
+    threads_[i].reset();
   }
+  threads_.clear();
+  closure_assistant_.reset();
+  sst_->on_deactivate();
+  sstm_->on_deactivate();
+  pose_pub_->on_deactivate();
+  return CallbackReturn::SUCCESS;
+}
 
+/*****************************************************************************/
+CallbackReturn SlamToolbox::on_cleanup(const rclcpp_lifecycle::State &)
+/*****************************************************************************/
+{
+  RCLCPP_INFO(get_logger(), "Cleaning up");
   smapper_.reset();
   dataset_.reset();
+  map_saver_.reset();
+  pose_helper_.reset();
+  laser_assistant_.reset();
+  scan_holder_.reset();
+  solver_.reset();
+
+  scan_filter_sub_.reset();
+  ssDesserialize_.reset();
+  ssSerialize_.reset();
+  ssPauseMeasurements_.reset();
+  ssMap_.reset();
+  sstm_.reset();
+  sst_.reset();
+  pose_pub_.reset();
+
+  tfB_.reset();
+  tfL_.reset();
+  tf_.reset();
+
+  lasers_.clear();
+  return CallbackReturn::SUCCESS;
+}
+
+/*****************************************************************************/
+CallbackReturn
+SlamToolbox::on_shutdown(const rclcpp_lifecycle::State & state)
```

Comment:
This is not what the shutdown transition is for in lifecycle nodes

Reply:
[612ffde29045c6170047881a9d70a3a77f5fe1d7]

- I checked nav2 functions and updated to log output only.

<https://github.com/SteveMacenski/slam_toolbox/blob/612ffde29045c6170047881a9d70a3a77f5fe1d7/src/slam_toolbox_common.cpp#L175-L182>